addons/l10n_it_edi_withholding/models/account_move.py

```diff
@@ -186,21 +189,57 @@ def _l10n_it_edi_get_extra_info(self, company, document_type, body_tree, incomin
                  + type_tax_use_domain),
                 vat_only=False)
             if pension_fund_tax:
-                pension_fund_taxes.append(pension_fund_tax)
+                pension_fund_taxes[vat_tax_factor_percent] = pension_fund_tax
             else:
                 message_to_log.append(Markup("%s<br/>%s") % (
                     _("Pension Fund tax not found"),
                     self.env['account.move']._compose_info_message(body_tree, '.'),
                 ))
         extra_info["pension_fund_taxes"] = pension_fund_taxes
 
+        # If the AssoSoftware specs are used on the invoice, then only apply
+        # the Pension Fund tax to the lines that show an AswCassPre
+        # additional tag (AltriDatiGestionali)
+        selector = ".//AltriDatiGestionali/TipoDato[contains(text(), 'AswCassPre')]"
+        if self.get_tag(body_tree, selector) is not None:
+            extra_info["pension_fund_assosoftware_tags"] = True
+
         return extra_info, message_to_log
 
+    def get_tag(self, element, selector):
+        if element is None:
+            return None
+        sub = element.xpath(selector)
+        return sub[0] if (sub is not None and sub) else None
+
+    def _get_pension_fund_tax_for_line(self, element, extra_info):
+        """ If there are AssoSoftware specific AltriDatiGestionale 'AswCassPre'
+            tags that specify which lines have pension funds, use them.
+            Otherwise apply the pension fund on all lines that have the related AliquotaIVA
+        """
+        pension_fund_map = extra_info.get('pension_fund_taxes', {})
+        if pension_fund_map and extra_info.get('pension_fund_assosoftware_tags'):
+            selector = ".//AltriDatiGestionali[TipoDato[contains(text(),'AswCassPre')]]/RiferimentoTesto"
+            if (reference_tag := self.get_tag(element, selector)) is not None:
+                pension_fund_taxes = pension_fund_map.values()
+                if match := re.match(r"(?P<kind>TC\d\d) \((?P<tax_rate>\d+)%\)", reference_tag.text):
+                    if pension_fund_tax := next((
+                        x for x in pension_fund_taxes
+                        if x.l10n_it_pension_fund_type == match['kind']
+                        and x.amount == float(match['tax_rate'])
```

Comment:
If you know the amount, why are you not getting it in O(1) from the getter of `pension_fund_map`?

Reply:
Will have a look Monday.

Reply:
Yes, actually, the mapping between VAT rate and pension fund can still be used.
I've changed the code a bit to make it more linear.